File: java/food-ordering/app/restate-app/src/main/java/dev/restate/sdk/examples/OrderWorkflow.java
package dev.restate.sdk.examples;

import static dev.restate.sdk.examples.utils.TypeUtils.statusToProto;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import dev.restate.sdk.RestateContext;
import dev.restate.sdk.common.CoreSerdes;
import dev.restate.sdk.common.TerminalException;
import dev.restate.sdk.examples.clients.PaymentClient;
import dev.restate.sdk.examples.clients.RestaurantClient;
import dev.restate.sdk.examples.generated.DeliveryManagerRestate;
import dev.restate.sdk.examples.generated.OrderProto.*;
import dev.restate.sdk.examples.generated.OrderStatusServiceRestate;
import dev.restate.sdk.examples.generated.OrderWorkflowRestate;
import dev.restate.sdk.examples.types.OrderRequest;
import dev.restate.sdk.examples.types.StatusEnum;
import java.time.Duration;
import java.util.UUID;

/**
 * Order processing workflow Gets called for each Kafka event that is published to the order topic.
 * The event contains the order ID and the raw JSON order. The workflow handles the payment, asks
 * the restaurant to start the preparation, and triggers the delivery.
 */
public class OrderWorkflow extends OrderWorkflowRestate.OrderWorkflowRestateImplBase {
  private final RestaurantClient restaurant = RestaurantClient.get();
  private final PaymentClient paymentClnt = PaymentClient.get();

  @Override
  public void handleOrderCreationEvent(RestateContext ctx, KafkaOrderEvent event)
      throws TerminalException {
    var orderStatusSend = OrderStatusServiceRestate.newClient(ctx);

    ObjectMapper mapper = new ObjectMapper();
    OrderRequest order;
    try {
      order = mapper.readValue(event.getPayload().toStringUtf8(), OrderRequest.class);
    } catch (JsonProcessingException e) {
      throw new TerminalException("Parsing raw JSON order failed: " + e.getMessage());
    }
    String id = order.getOrderId();

    // 1. Set status
    orderStatusSend.oneWay().setStatus(statusToProto(id, StatusEnum.CREATED));

    // 2. Handle payment
    String token = ctx.sideEffect(CoreSerdes.STRING_UTF8, () -> UUID.randomUUID().toString());
    boolean paid =
        ctx.sideEffect(
            CoreSerdes.BOOLEAN, () -> paymentClnt.charge(id, token, order.getTotalCost()));

    if (!paid) {
      orderStatusSend.oneWay().setStatus(statusToProto(id, StatusEnum.REJECTED));
      return;
    }

    // 3. Schedule preparation
    orderStatusSend.setStatus(statusToProto(order.getOrderId(), StatusEnum.SCHEDULED));
    ctx.sleep(Duration.ofMillis(order.getDeliveryDelay()));

    // 4. Trigger preparation
    var preparationAwakeable = ctx.awakeable(CoreSerdes.VOID);
    ctx.sideEffect(() -> restaurant.prepare(id, preparationAwakeable.id()));
    orderStatusSend.setStatus(statusToProto(id, StatusEnum.IN_PREPARATION));

    preparationAwakeable.await();
    orderStatusSend.setStatus(statusToProto(id, StatusEnum.SCHEDULING_DELIVERY));

    // 5. Find a driver and start delivery
    var deliveryAwakeable = ctx.awakeable(CoreSerdes.VOID);

    var deliveryRequest =
        DeliveryRequest.newBuilder()
            .setOrderId(id)
            .setRestaurantId(order.getRestaurantId())
            .setCallback(deliveryAwakeable.id())
            .build();
    DeliveryManagerRestate.newClient(ctx).oneWay().start(deliveryRequest);
    deliveryAwakeable.await();
    orderStatusSend.setStatus(statusToProto(order.getOrderId(), StatusEnum.DELIVERED));
  }
}

File: typescript/food-ordering/app/src/order-app/order_workflow.ts
import * as restate from "@restatedev/restate-sdk";
import * as deliveryManager from "../delivery-app/delivery_manager";
import {Order, Status} from "./types/types";
import * as orderstatus from "./order_status_service";
import {getPaymentClient} from "./clients/payment_client";
import {getRestaurantClient} from "./clients/restaurant_client";
import {fail} from "./utils/utils";

export const service: restate.ServiceApi<typeof router> = { path: "order-workflow" };

const restaurant = getRestaurantClient();
const paymentClnt = getPaymentClient();

export const router = restate.keyedRouter({
  create: async ( ctx: restate.RpcContext, orderId: string, order: Order) => {
    const { id, totalCost, deliveryDelay } = order;

    // 1. Set status
    ctx.send(orderstatus.service).setStatus(id, Status.CREATED);

    // 2. Handle payment
    const token = ctx.rand.uuidv4();
    // console.info(token);
    const paid = await ctx.sideEffect(async () => paymentClnt.charge(id, token, totalCost));

    if (!paid) {
      ctx.send(orderstatus.service).setStatus(id, Status.REJECTED);
      return;
    }

    // 3. Wait for desired preparation time
    ctx.send(orderstatus.service).setStatus(id, Status.SCHEDULED);
    await ctx.sleep(deliveryDelay);

    // 4. Trigger preparation
    const preparationPromise = ctx.awakeable();
    await ctx.sideEffect(() => restaurant.prepare(id, preparationPromise.id));
    ctx.send(orderstatus.service).setStatus(id, Status.IN_PREPARATION);

    await preparationPromise.promise;
    ctx.send(orderstatus.service).setStatus(id, Status.SCHEDULING_DELIVERY);

    // fail(orderId, "Failed");
    // 5. Find a driver and start delivery
    await deliver(ctx, order);
    ctx.send(orderstatus.service).setStatus(id, Status.DELIVERED);
    //
  }
});

async function deliver(ctx: restate.RpcContext, order: Order) {
  const deliveryId = ctx.rand.uuidv4();
  const deliveryPromise = ctx.awakeable<void>();
  await ctx.rpc(deliveryManager.service).start(deliveryId, {order, promise: deliveryPromise.id})
  await deliveryPromise.promise;
}
